# AI ANALYST LAB — Python Crash Course

![](../_img/ai_analyst_lab.png)

### From Zero to Python — a gentle companion for people who have never programmed
## Session 6: pandas — grouping & summarising

Feedback should be sent to [goran.milovanovic@datakolektiv.com](mailto:goran.milovanovic@datakolektiv.com).

Last session you loaded 891 real passengers into a table and learned to look one **slice** at a time — *"keep only the survivors."* Today you learn the move a data analyst reaches for more than any other: **split the table into groups, and get one number per group — all at once.**

This is where a single line answers a real question: *"what was the survival rate for women versus men?"* — and the famous story of the Titanic falls straight out of the data.

Go slowly. You **cannot break anything**.

***
### What we will do today

- **Get set up again** — a fresh file, in the same folder, with the data re-loaded (step by step).
- Take a quick **headcount** of any column with **`.value_counts()`**.
- Meet **`.groupby()`** — split into groups, summarise each, all in one line.
- Discover why the average of a **0/1 column** is simply a **rate**.
- Group by **two columns at once** (sex *and* class) for the deeper story.
- **Sort** the answer so it reads top to bottom with **`.sort_values()`**.

Work in the **IDLE editor**, running with **Run → Run Module** each time. *(On a Mac, use the menu — F5 opens Dictation.)*

***
## 1. Getting set up again

Every time you close Python, it **forgets everything** — every variable, and the whole table you loaded. So each new session begins by **re-loading the data**. That is completely normal; it is why the *first* few lines of nearly every analysis file look the same.

We'll do it in **two small steps**.

### Step 1 — A fresh file, in the same folder as before

Open **IDLE → File → New File**, and **save it** (File → Save) **into the same `AI_AnalystLAB` folder** you made in Session 5 — the folder where **`Titanic_train.csv` already lives**. Give this one the name **`grouping.py`**.

Keeping today's file in that same folder means the data file is right there beside it, exactly as before.

### Step 2 — Load the data again

Type these two lines at the top of `grouping.py` and run it (**Run → Run Module**). They are the same two lines from last session — bring in pandas, and read the table:

In [1]:
import pandas as pd

passengers = pd.read_csv("Titanic_train.csv")
print("Loaded:", passengers.shape)

Loaded: (891, 12)


You should see **`Loaded: (891, 12)`** — the full table is back, ready to work with.

> **⚠️ Same one error to watch for.** If you get **`FileNotFoundError`** for `'Titanic_train.csv'`, the data file isn't in the same folder as `grouping.py`. You downloaded it in **Session 5, Step 3** — make sure it sits beside this file. If it's gone, download it again from Session 5, or ask your instructor.

***
## 2. The question that filtering can't answer neatly

Last session you kept **one** group at a time — *the survivors* — with a mask. But real questions compare **many** groups at once: *how did survival differ between **women and men**? Between **first, second, and third class**?*

You *could* do it by hand: filter the women, compute their rate; then filter the men, compute theirs; then write both down. Tedious — and it gets worse the more groups there are. pandas has a single tool that does **all the groups in one stroke**. But first, a gentle warm-up.

***
## 3. A quick headcount — `.value_counts()`

Before summarising, analysts often just ask *"how many of each?"* — a plain **headcount** of a column. That's **`.value_counts()`**: hand it a column, and it tallies how many times each value appears — like sorting a bag of coloured beads into piles and counting each pile.

In [2]:
print(passengers["Sex"].value_counts())

Sex
male      577
female    314
Name: count, dtype: int64


Out come the tallies: **577 male, 314 female**. Try it on ticket class too:

In [3]:
print(passengers["Pclass"].value_counts())

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


**491** passengers in 3rd class, **216** in 1st, **184** in 2nd.

> Notice the order: `.value_counts()` lists the **most common first** (3rd class, then 1st, then 2nd) — not `1, 2, 3`. It sorts by *how many*, not by the value itself. Good to know so the order never surprises you.

***
## 4. The main event — `.groupby()`

Here is the analyst's workhorse. Picture a teacher with a big pile of exam papers: they **split** the papers into stacks — one stack per class — then work out the **average score for each stack**, and finally **combine** those averages into a short list. **`.groupby()`** does exactly that, in one line: *split* the table into groups, *summarise* each group, *combine* the answers into a small table.

Let's ask the real question — the survival **rate** for each sex:

In [4]:
survival_by_sex = passengers.groupby("Sex")["Survived"].mean()
print(survival_by_sex.round(3))

Sex
female    0.742
male      0.189
Name: Survived, dtype: float64


Read the line left to right, like a sentence:

- **`passengers.groupby("Sex")`** — split the table into groups by the `Sex` column (one group of women, one of men);
- **`["Survived"]`** — in each group, look at the `Survived` column;
- **`.mean()`** — take its average within each group;
- **`.round(3)`** — tidy the answer to 3 decimals so it's easy to read.

The result: **women `0.742`, men `0.189`** — about **74%** of the women survived, against only **19%** of the men. In one line, the Titanic's grim *"women and children first"* is written in the data.

> **Why is an *average* a *rate*?** The `Survived` column is all **1**s (survived) and **0**s (did not). If you average a pile of 1s and 0s, the answer *is* just the **fraction that are 1** — so the **mean of `Survived` is the survival rate**. That small trick — averaging a yes/no column to get a proportion — is used constantly in the main course.

***
## 5. Grouping by two things at once

You can split by **more than one column** — just hand `groupby` a **list** of column names inside `[ ]`. Let's see survival by **sex *and* class** together:

In [5]:
print(passengers.groupby(["Sex", "Pclass"])["Survived"].mean().round(3))

Sex     Pclass
female  1         0.968
        2         0.921
        3         0.500
male    1         0.369
        2         0.157
        3         0.135
Name: Survived, dtype: float64


Now the story deepens. Read down the six groups:

- **First-class women: `0.968`** — almost everyone survived.
- **Third-class men: `0.135`** — almost no one did.

Sex mattered enormously, but so did class — and the two together tell a sharper story than either alone. That is the whole point of grouping: **let the data split itself and compare.**

*(The `["Sex", "Pclass"]` is a **list** of two columns — the same square-bracket list you met in Session 2 — telling `groupby` to make a group for every sex-and-class combination.)*

***
## 6. Ordering the answer — `.sort_values()`

A summary is easier to read when it's **sorted**. **`.sort_values()`** puts the numbers in order — smallest to largest by default, or largest first if you ask. Let's rank the classes by survival, highest first:

In [6]:
survival_by_class = passengers.groupby("Pclass")["Survived"].mean()
print(survival_by_class.sort_values(ascending=False).round(3))

Pclass
1    0.630
2    0.473
3    0.242
Name: Survived, dtype: float64


Top to bottom: **1st class `0.630`, 2nd `0.473`, 3rd `0.242`** — survival fell steadily as you moved down the classes. The **`ascending=False`** part simply means *"biggest first"*; leave it out and you get smallest first.

Sorting doesn't change the numbers — it just arranges them so the story reads in order, which matters even more once you start drawing charts of them.

***
## What you can now do — and what's next

Today you learned the summarising verbs at the centre of everyday analysis:

- take a quick **headcount** with **`.value_counts()`**;
- **`.groupby()`** — split into groups and summarise each, all in one line;
- read the average of a **0/1 column** as a **rate**;
- group by **two columns** for a richer breakdown;
- and **`.sort_values()`** to order the result.

With Sessions 5 and 6 you now have the core of pandas: **load and look → select and filter → group and summarise.** That is genuinely most of what the analyses in the main AI Analyst LAB course do to their data.

Next, **Session 7** turns numbers into **pictures**. You'll draw the very survival rates you just computed as **bar charts**, and a passenger-age **histogram** — because a chart often tells the story faster than a table.

> **⚠️ Quick reminders:**
> - Python **forgets between runs** — re-load the data at the top of every new file.
> - Keep the `.py` file and `Titanic_train.csv` in the **same folder** (`FileNotFoundError` means they aren't).
> - Read a `groupby` line as a sentence: **group by → pick a column → summarise**.

See you in Session 7.

<hr>

![](../_img/DK_Logo_White_150.png)

DataKolektiv, 2026.

[hello@datakolektiv.com](mailto:hello@datakolektiv.com)

<font size=1>License: [GPLv3](../LICENSE). This Notebook is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This Notebook is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</font>